### PR TITLE
commit for acl testing

### PR DESCRIPTION
**Pull request by huanwwwq**

### CODE REVIEW

File: pike/smb2.py [hunk 1]
```diff
@@ -1430,6 +1430,141 @@ def _get_offset(self,owner_sid=None,group_sid=None,sacl=None,dacl=None):
...
+        """
+        decode the raw data to the specified end of buffer
+        """
+        self.revision = cur.decode_uint8le()
+        self.sub_authority_count = cur.decode_uint8le()
+        id_auth_high = cur.decode_uint16be()
+        id_auth_low = cur.decode_uint32be()
+        self.identifier_authority = (id_auth_high << 32) + id_auth_low
+        for i in range(self.sub_authority_count):
+            self.sub_authority += [cur.decode_uint32le()]
```
Comment (masenf) [inline pike/smb2.py]:
```
self.sub_authority.append(cur.decode_uint32le())
```

File: pike/smb2.py [hunk 2]
```diff
@@ -1984,6 +2123,75 @@ class FileInformation(Information):
...
+            self.other = cur.decode_bytes(self.end - cur)
+
+    def _encode(self, cur):
+        cur.encode_uint8le(self.revision)
+        cur.encode_uint8le(0)
+        cur.encode_uint16le(self.control)
+        cur.encode_uint32le(self.offset_owner)
+        cur.encode_uint32le(self.offset_group)
+        cur.encode_uint32le(self.offset_sacl)
+        cur.encode_uint32le(self.offset_dacl)
```
Comment (masenf) [inline pike/smb2.py]: you need to create `hole`s for these values and then fill them in later after you have encoded the previous values into the buffer and _know_ what the offset should be. See the `_encode` routine for `SessionSetupRequest`:

https://github.com/emc-isilon/pike/blob/master/pike/smb2.py#L725-L735

File: pike/smb2.py (hunk 1 above)
Comment (huanwwwq) [reply pike/smb2.py]: accept and fixed

File: pike/smb2.py (hunk 2 above)
Comment (huanwwwq) [reply pike/smb2.py]: accept and fixed

**Comment by masenf** [comment]
Something you might want to consider is defining the ace type and ace flags enum as well as casting the access mask as an `Access`. This will give you symbolic names to use in test code and richer output when printing the structs.

With the code here, the struct looks like this

```
FileSecurityInformation
  raw_data: 
  revision: 1
  sbz1: 0
  control: 32772
  offset_owner: 20
  offset_group: 36
  offset_sacl: 0
  offset_dacl: 52
  other: 0x
  owner_sid: NT_SID
    raw_data: 
    revision: 1
    sub_authority_count: 2
    identifier_authority: 22
    sub_authority: [1, 0]
  group_sid: NT_SID
    raw_data: 
    revision: 1
    sub_authority_count: 2
    identifier_authority: 22
    sub_authority: [2, 0]
  sacl: 
  dacl: NT_ACL
    acl_revision: 2
    sbz1: 0
    acl_size: 32
    ace_count: 1
    sbz2: 0
    raw_data: 
    NT_ACE
      ace_type: 0
      ace_flags: 0
      ace_size: 24
      access_mask: 2032127
      sid: NT_SID
        raw_data: 
        revision: 1
        sub_authority_count: 2
        identifier_authority: 22
        sub_authority: [1, 0]
      raw_data: 
```

You may also consider overwriting `__str__` for `NT_SID` to display the formatted string instead of the raw structure.

**Comment by huanwwwq** [comment]
> Something you might want to consider is defining the ace type and ace flags enum as well as casting the access mask as an `Access`. This will give you symbolic names to use in test code and richer output when printing the structs.
> 
> With the code here, the struct looks like this
> 
> ```
> FileSecurityInformation
>   raw_data: 
>   revision: 1
>   sbz1: 0
>   control: 32772
>   offset_owner: 20
>   offset_group: 36
>   offset_sacl: 0
>   offset_dacl: 52
>   other: 0x
>   owner_sid: NT_SID
>     raw_data: 
>     revision: 1
>     sub_authority_count: 2
>     identifier_authority: 22
>     sub_authority: [1, 0]
>   group_sid: NT_SID
>     raw_data: 
>     revision: 1
>     sub_authority_count: 2
>     identifier_authority: 22
>     sub_authority: [2, 0]
>   sacl: 
>   dacl: NT_ACL
>     acl_revision: 2
>     sbz1: 0
>     acl_size: 32
>     ace_count: 1
>     sbz2: 0
>     raw_data: 
>     NT_ACE
>       ace_type: 0
>       ace_flags: 0
>       ace_size: 24
>       access_mask: 2032127
>       sid: NT_SID
>         raw_data: 
>         revision: 1
>         sub_authority_count: 2
>         identifier_authority: 22
>         sub_authority: [1, 0]
>       raw_data: 
> ```
> 
> ~You may also consider overwriting `__str__` for `NT_SID` to display the formatted string instead of the raw structure.~ Nevermind, i see you've already done this, but it's not working.
> 
> Ah okay, turns out that you need to override `_str(self, indent)` (see pike/core.py Frame._str)

accept and fixed
FileSecurityInformation
  raw_data:
  revision: 1
  sbz1: 0
  control: 33796
  offset_owner: 20
  offset_group: 36
  offset_sacl: 0
  offset_dacl: 64
  other: 0x
  owner_sid: S-1-5-32-544
  group_sid: S-1-5-21-1714174032-3688233909-525558184-513
  sacl:
  dacl: NT_ACL
    acl_revision: ACL_REVISION
    sbz1: 0
    acl_size: 88
    ace_count: 3
    sbz2: 0
    raw_data:
    NT_ACE
      ace_type: ACCESS_ALLOWED_ACE_TYPE
      ace_flags: INHERITED_ACE
      ace_size: 36
      access_mask: FILE_ADD_FILE | FILE_READ_DATA | FILE_TRAVERSE | SYNCHRONIZE | WRITE_OWNER | FILE_WRITE_DATA | FILE_WRITE_EA | FILE_EXECUTE | FILE_READ_EA | WRITE_DAC | FILE_DELETE_CHILD | FILE_ADD_SUBDIRECTORY | READ_CONTROL | FILE_LIST_DIRECTORY | FILE_READ_ATTRIBUTES | FILE_WRITE_ATTRIBUTES | FILE_APPEND_DATA | DELETE
      sid: S-1-5-21-1714174032-3688233909-525558184-500
      raw_data:
    NT_ACE
      ace_type: ACCESS_ALLOWED_ACE_TYPE
      ace_flags: INHERITED_ACE
      ace_size: 24
      access_mask: FILE_ADD_FILE | FILE_READ_DATA | FILE_TRAVERSE | SYNCHRONIZE | WRITE_OWNER | FILE_WRITE_DATA | FILE_WRITE_EA | FILE_EXECUTE | FILE_READ_EA | WRITE_DAC | FILE_DELETE_CHILD | FILE_ADD_SUBDIRECTORY | READ_CONTROL | FILE_LIST_DIRECTORY | FILE_READ_ATTRIBUTES | FILE_WRITE_ATTRIBUTES | FILE_APPEND_DATA | DELETE
      sid: S-1-5-32-544
      raw_data:
    NT_ACE
      ace_type: ACCESS_ALLOWED_ACE_TYPE
      ace_flags: INHERITED_ACE
      ace_size: 20
      access_mask: FILE_ADD_FILE | FILE_READ_DATA | FILE_TRAVERSE | SYNCHRONIZE | WRITE_OWNER | FILE_WRITE_DATA | FILE_WRITE_EA | FILE_EXECUTE | FILE_READ_EA | WRITE_DAC | FILE_DELETE_CHILD | FILE_ADD_SUBDIRECTORY | READ_CONTROL | FILE_LIST_DIRECTORY | FILE_READ_ATTRIBUTES | FILE_WRITE_ATTRIBUTES | FILE_APPEND_DATA | DELETE
      sid: S-1-5-18
      raw_data:

**Comment by huanwwwq** [comment]
> Retract my previous approval.
> 
> I went to go off and try to use this code and found that it worked fine for reading, but attempting to modify a security descriptor was cumbersome.
> 
> The following fields also need to have `hole`s and/or dynamically calculate the value on encode:
> 
> NT_ACL.acl_size - byte count length of serialized ACL - use a hole
> NT_ACL.ace_count - should be the number of children present - may encode directly
> 
> NT_ACE.ace_size - byte count of serialized ACE - use a hole
> 
> NT_SID.sub_authority_count - should be the length of the sub_authority list
> 
> There's a tradeoff you'll need to make: it should be possible to bypass the calculated values (for instance if the user wants to create an invalid structure), but it should also be easy to edit an existing security descriptor.
> 
> My advice here is to follow the pattern you have in `FileSecurityInformation` where if the value is 0, then you replace it with a calculated value (however you may consider a different sentinel, like `None`, because 0 is a real value). HOWEVER, for the fields that will cause changes to the calculated offsets when modified, i think you should make those `@property` and reset the offset/size value to the sentinel when these fields are modified.

accept and fixed

add ACE to ACL is a little bit ugly.

    @property
    def aces(self):
        return self._entries

    @aces.setter
    def aces(self, ace_list):
        if not isinstance(ace_list, list):
            raise ValueError('aces must be a list!')
        self._entries = ace_list
        self.acl_size = -1
        self.ace_count = -1

have to set the aces as a whole, can not append new item

**Comment by huanwwwq** [comment]
add new commits to allow acl.aces.append(newace)

**Comment by masenf** [comment]
I also noticed that the sid isn't showing up once the field was changed to start with an underscore?

```
FileSecurityInformation
  raw_data: 
  revision: 1
  sbz1: 0
  control: PS | OD
  offset_owner: 20
  offset_group: 36
  offset_sacl: 0
  offset_dacl: 52
  other: 0x
  NT_ACL
    acl_revision: ACL_REVISION
    sbz1: 0
    acl_size: 32
    ace_count: 1
    sbz2: 0
    raw_data: 
    NT_ACE
      ace_type: ACCESS_ALLOWED_ACE_TYPE
      ace_flags: 0
      ace_size: 24
      access_mask: FILE_ADD_FILE | FILE_READ_DATA | FILE_TRAVERSE | SYNCHRONIZE | WRITE_OWNER | FILE_WRITE_DATA | FILE_WRITE_EA | FILE_EXECUTE | FILE_READ_EA | WRITE_DAC | FILE_DELETE_CHILD | FILE_ADD_SUBDIRECTORY | READ_CONTROL | FILE_LIST_DIRECTORY | FILE_READ_ATTRIBUTES | FILE_WRITE_ATTRIBUTES | FILE_APPEND_DATA | DELETE
      raw_data: 
```

none of the underscore fields are showing up =/

**Comment by masenf** [comment]
i think my getter setter advice to update the calculated values was a bad advice. you'll have to set those fields back to regular, but I have a diff that i was messing with, which i'll post=

**Comment by isi-mfurer** [comment]
sorry for using multiple accounts, i didn't realize i was logged in with my personal account.

anyway, here are my code suggestions for you: https://github.com/huanwwwq/pike/pull/1

feel free to modify that or change things around. i had some ideas for the implementation, but the last 2 times i made suggestions, the interface didn't really end up working out. so this suggestion, I actually coded up to make sure that it worked as expected.

additionally, i discovered why the tests were working against windows 2016 and failing against onefs: windows was ignoring the contents of the request because the security info flags were not getting set.


so, now that that's fixed (in my PR to you), you have some work ahead of you looking at pcaps and figuring out why it doesn't work! good luck.

**Pull request closed without merging**